base/strings/annotated.jl

```diff
@@ -386,3 +386,112 @@ annotations(s::SubString{<:AnnotatedString}, pos::UnitRange{<:Integer}) =
 Get all annotations of `chr`.
 """
 annotations(c::AnnotatedChar) = c.annotations
+
+## AnnotatedIOBuffer
+
+"""
+    AnnotatedIOBuffer <: IO
+
+A thin wrapper around an [`IOBuffer`](@ref) that also preserves annotations (see [`AnnotatedString`](@ref)).
+
+In addition to supporting the usual [`IOBuffer`](@ref) methods (e.g.
+[`position`](@ref), [`seek`](@ref), [`truncate`](@ref)), `AnnotatedIOBuffer`
+also supports a specialised [`read(anniobuf, AnnotatedString)`](@ref) method for
+reading out both content and accumulated annotations. This method will read from
+the start of the buffer when `eof(anniobuf)`.
+
+Note that `AnnotatedIOBuffer` does *not* have the `:color` IO property.
+
+# Examples
+
+```jldoctest
+julia> aio = Base.AnnotatedIOBuffer()
+Base.AnnotatedIOBuffer(0 bytes, 0 annotations)
+
+julia> print(aio, "hello ", Base.AnnotatedString("world", [(1:5, :i => 2)]))
+
+julia> read(aio, Base.AnnotatedString)
+"hello world"
+
+julia> read(aio, Base.AnnotatedString) |> Base.annotations
+1-element Vector{Tuple{UnitRange{Int64}, Pair{Symbol, Any}}}:
+ (7:11, :i => 2)
+```
+"""
+struct AnnotatedIOBuffer <: IO
+    io::IOBuffer
+    annotations::Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}
+end
+
+AnnotatedIOBuffer(io::IOBuffer) = AnnotatedIOBuffer(io, Vector{Tuple{UnitRange{Int}, Pair{Symbol, Any}}}())
+AnnotatedIOBuffer() = AnnotatedIOBuffer(IOBuffer())
+
+function show(io::IO, aio::AnnotatedIOBuffer)
+    show(io, AnnotatedIOBuffer)
+    print(io, '(', aio.io.size, " byte", ifelse(aio.io.size == 1, "", "s"), ", ",
+          length(aio.annotations), " annotation", ifelse(length(aio.annotations) == 1, "", "s"), ")")
+end
+
+isopen(io::AnnotatedIOBuffer) = isopen(io.io)
+close(io::AnnotatedIOBuffer) = close(io.io)
+eof(io::AnnotatedIOBuffer) = eof(io.io)
+peek(io::AnnotatedIOBuffer) = peek(io.io)
+position(io::AnnotatedIOBuffer) = position(io.io)
+seek(io::AnnotatedIOBuffer, n::Integer) = (seek(io.io, n); io)
+skip(io::AnnotatedIOBuffer, n::Integer) = (skip(io.io, n); io)
+mark(io::AnnotatedIOBuffer) = mark(io.io)
+reset(io::AnnotatedIOBuffer) = reset(io.io)
+unmark(io::AnnotatedIOBuffer) = unmark(io.io)
+ismarked(io::AnnotatedIOBuffer) = ismarked(io.io)
+copy(io::AnnotatedIOBuffer) = AnnotatedIOBuffer(copy(io.io), copy(io.annotations))
+
+function write(io::AnnotatedIOBuffer, astr::Union{AnnotatedString, SubString{<:AnnotatedString}})
+    astr = AnnotatedString(astr)
+    offset = position(io.io)
+    for (region, annot) in astr.annotations
+        start, stop = first(region), last(region)
+        push!(io.annotations, (start+offset:stop+offset, annot))
+    end
+    write(io.io, String(astr))
+end
+write(io::AnnotatedIOBuffer, c::AnnotatedChar) = write(io, AnnotatedString(c))
+write(io::AnnotatedIOBuffer, x::AbstractString) = write(io.io, x)
+write(io::AnnotatedIOBuffer, s::Union{SubString{String}, String}) = write(io.io, s)
+write(io::AnnotatedIOBuffer, b::UInt8) = write(io.io, b)
```

Comment:
```suggestion
write(io::AnnotatedIOBuffer, b::UInt8) = write(io.io, b)
unsafe_write(io::AnnotatedIOBuffer, b::Ptr{UInt8}, len::UInt) = unsafe_write(io.io, b, len)
```

Performance will typically be punishingly slow, unless you defined `unsafe_write` and `unsafe_read`. You could instead make this a subtype of `AbstractPipe` instead of directly implementing all of `IO`, since then you only need to define `pipe_reader` and `pipe_writer` and methods that you want to override (e.g. this is how `IOContext` is implemented over `IOBuffer`, to add/override `get` but inherit everything else)

Reply:
I've now added `unsafe_write` and `unsafe_read` passthrough methods. I've stayed away from `AbstractPipe` because of uncertainty around #51808 and it just generally not being clear what exactly the `AbstractPipe` interface is (what do you need to do? which methods are then provided?).

From looking at the `AbstractPipe` code, it seems like you don't get `seek`, `skip`, or `position` (which I'd think are useful enough to be worth having), so we could maybe subtype that and then just provide those methods? I wonder if it would make more sense to do so after the `AbstractPipe` interface is better documented though.

Reply:
Yeah, we could consider adding `seek`, `skip`, and `position` that affects the `pipe_reader`. Adding to the docstring would also be helpful. Originally it was basically just an internal helper type because it was a huge annoyance to need to define all those methods you just wrote above and then need to write tests for all of them. But clearly it seems to be pretty broadly useful too for defining other new IO wrappers like this.

Reply:
Looking again in comparison to AbstractPipe inheritence, you also may be missing forwarding opportunities for:
isreadable, iswritable, buffer_writes, flush, readavailable, bytesavailable, reseteof, copyuntil, readuntil_vector!, readbytes!, peek, wait_readnb, wait_close

Many of these will have generic fallbacks, though they may have less functionality, but are probably fine for IOBuffer as there may not be a specialized version for IOBuffer anyways

Reply:
Ok, at this point I'll just switch to inheriting from `AbstractPipe`.

A subsequent PR from somebody who has more of a sense of the purpose of it adding/clarifying docs explaining `AbstractPipe` a bit more I think would be well worth it though.

Reply:
I just merged more docs for that if you look now :)

Reply:
Oh brilliant! Thanks for that Jameson :)